cfb: support point drawing non vtiled displays

yishai1999 opened the pull request:
Added support for drawing points on non-vtiled displays.

yishai1999 committed:
cfb: support non vtiled displays

Added support for drawing points on non-vtiled displays.

Signed-off-by: Yishai Jaffe <yishai1999@gmail.com>

soburi reviewed (changes requested):
Please also add test cases.

yishai1999 commented:
> Please also add test cases.

this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?

soburi commented:
@yishai1999 

> this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?

Yes, that's right.
It also means that you will need to extend the SDL emulator if necessary.

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_sdl.c

yishai1999 commented:
> @yishai1999
> 
> > this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?
> 
> Yes, that's right. It also means that you will need to extend the SDL emulator if necessary.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_sdl.c

The problem is that the SDL emulator is `SCREEN_INFO_MONO_VTILED`. It can't be both VTILED and not VTILED. How do you propose we solve this conflict?

yishai1999 committed:
tests: display: cfb: added htiled display test cases

Added the option to make the SDL emulator htiled and added test cases to
test it.

Signed-off-by: Yishai Jaffe <yishai1999@gmail.com>

yishai1999 commented:
> @yishai1999
> 
> > this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?
> 
> Yes, that's right. It also means that you will need to extend the SDL emulator if necessary.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_sdl.c


How does this look?

soburi commented:
> > @yishai1999
> > > this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?
> > 
> > 
> > Yes, that's right. It also means that you will need to extend the SDL emulator if necessary.
> > https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_sdl.c
> 
> How does this look?

There's nothing wrong with going in this direction. Next, it looks like we need to implement drawing processing in the following places.

https://github.com/zephyrproject-rtos/zephyr/blob/f6d0e7d97c03e4ba090ff7e44c25dc3dd373e572/drivers/display/display_sdl.c#L182-L224

This also requires testing, so it might be a good idea to separate it into another independent PR.

yishai1999 commented:
> > > @yishai1999
> > > > this would require a device that is not `SCREEN_INFO_MONO_VTILED`. Am I mistaken?
> > > 
> > > 
> > > Yes, that's right. It also means that you will need to extend the SDL emulator if necessary.
> > > https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_sdl.c
> > 
> > 
> > How does this look?
> 
> There's nothing wrong with going in this direction. Next, it looks like we need to implement drawing processing in the following places.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f6d0e7d97c03e4ba090ff7e44c25dc3dd373e572/drivers/display/display_sdl.c#L182-L224
> 
> This also requires testing, so it might be a good idea to separate it into another independent PR.

So approve?

I'll try look at the write function as well

yishai1999 commented:
@soburi can you approve?

soburi commented:
> @soburi can you approve?

I think we also need to update this part.

https://github.com/zephyrproject-rtos/zephyr/blob/d8edd781841f6a68076036d232f9a67fea090ab1/drivers/display/display_sdl.c#L205-L223

yishai1999 commented:
> One more thing.
> 
> Please add support for the display test.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/6efca1a9ff9a7bcbf742d25fc213add25eb7e34c/tests/drivers/display/display_read_write/src/main.c#L91-L100

Done

soburi commented on tests/drivers/display/display_read_write/src/main.c:
```diff
@@ -142,7 +148,7 @@ ZTEST(display_read_write, test_clear)
 	uint8_t data[4] = {0xFA, 0xAF, 0x9F, 0xFA};
-	uint8_t height = (is_tiled ? 8 : 1);
+	uint8_t height = (is_mono ? 8 : 1);
 	uint16_t width = sizeof(data) / bpp;
```
I think the width value also needs to be linked to is_vtiled.

soburi commented on tests/drivers/display/display_read_write/src/main.c:
```diff
@@ -22,7 +22,8 @@ static uint8_t disp_buffer[DT_PROP(DT_CHOSEN(zephyr_display), width) *
 static uint8_t bpp;
-static bool is_tiled;
+static bool is_vtiled;
+static bool is_mono;
```
Since this is a complement relationship,
I think it would be better to write the processing symmetrically using is_vtiled and is_htiled.

The part where it says is_mono can be replaced with the following.

```
if (is_vtiled || is_htiled) {
// mono case
}
```

yishai1999 commented:
Fixed CR comments

soburi commented:
It seems something is wrong...

```
$ ./twister-out/native_sim_native_64/host/tests/drivers/display/display_read_write/drivers.display.read_write.sdl.mono10/zephyr/zephyr.exe
*** Booting Zephyr OS build v4.1.0-467-g218344d1caaf ***
Running TESTSUITE display_read_write
===================================================================
START - test_clear
ASSERTION FAIL [(desc->pitch * desc->height) <= (desc->buf_size * 8U)] @ WEST_TOPDIR/zephyr/drivers/display/display_sdl.c:387
        Read buffer is too small
@ WEST_TOPDIR/zephyr/lib/os/assert.c:43
[00:00:00.000,000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00:00:00.000,000] <err> os: Current thread: 0x4633a0 (test_clear)
[00:00:00.000,000] <err> os: Halting system
Exiting due to fatal error
```

soburi commented:
Please split out commits into 3 parts or sorting commits.

`cfb: shell: fix draw point help`
-> Split into a separate PR. This can be accepted quickly.

`drivers: display: sdl: add non-vtiled mono write`
`tests: drivers: display: add support for htiled mode` 
-> Create new PR to aggregate SDL-display-related changes.

yishai1999 commented:
> Please split out commits into 3 parts or sorting commits.
> 
> `cfb: shell: fix draw point help` -> Split into a separate PR. This can be accepted quickly.
> 
> `drivers: display: sdl: add non-vtiled mono write` `tests: drivers: display: add support for htiled mode` -> Create new PR to aggregate SDL-display-related changes.

Shouldn't `test: display: added htiled display test cases` also be in this new PR?

soburi commented:
@yishai1999 

> Shouldn't test: display: added htiled display test cases also be in this new PR?

I think it's a good idea to separate the display changes and the associated test additions into separate PRs.
The CFB changes can only be made after this is complete.

yishai1999 commented:
> @yishai1999
> 
> > Shouldn't test: display: added htiled display test cases also be in this new PR?
> 
> I think it's a good idea to separate the display changes and the associated test additions into separate PRs. The CFB changes can only be made after this is complete.

I opened https://github.com/zephyrproject-rtos/zephyr/pull/88126

yishai1999 commented:
@soburi Can this be merged too?

soburi commented:
@yishai1999 

This test must be modified as it only considers the vtiled format.

https://github.com/zephyrproject-rtos/zephyr/blob/b3ea4f8d5c9b39d43b4d074b4e25bbc7a14e6e2d/tests/subsys/display/cfb/basic/src/utils.c#L22-L29

yishai1999 commented:
@soburi Would you be able to help?

soburi reviewed (approved):
Ideally, I would like to confirm that the test works, but since #88714 is still not expected to be fixed and the content is probably not a problem, I will approve it.

VynDragon reviewed (approved):
LGTM
